ouroboros-network/src/Ouroboros/Network/PeerSelection/KnownPeers.hs

```diff
@@ -105,26 +99,44 @@ data KnownPeerInfo = KnownPeerInfo {
        -- thus it can be present for warm or cold peers.  It's purpose is to
        -- provide information to demotion policies.
        --
-       knownPeerTepid     :: !Bool
+       -- It is also used as useful information for the Peer Selection Governor
+       -- when deciding which peers to share when Peer Sharing.
+       --
+       knownPeerTepid     :: !Bool,
+
+       -- | Indicates current remote Peer Willingness information.
+       --
+       -- If a connection with this address hasn't been established we won't
+       -- have any information about this particular flag
+       --
+       -- It is used by the Peer Sharing logic to decide if we should share/ask
+       -- about/to this peer's address to others.
+       knownPeerSharing   :: !(Maybe PeerSharing),
+
+       -- | Indicates current local Peer Willingness information.
+       --
+       -- If this address didn't come from a local configuration then this
+       -- value is set to 'DoAdvertise' by default.
+       --
+       -- It is used by the Peer Sharing logic to decide if we should share
+       -- about this peer's address to others.
+       knownPeerAdvertise :: !PeerAdvertise,
+
+       -- | Indicates if peer came from ledger.
+       --
+       -- It is used so we can filter out the ledger Peers from a Peer Sharing
+       -- reply, since ledger peers are not particularly what one is looking for
+       -- in a Peer Sharing reply.
+       --
+       knownLedgerPeer    :: !LedgerPeer
      }
   deriving (Eq, Show)
 
 
 invariant :: Ord peeraddr => KnownPeers peeraddr -> Bool
 invariant KnownPeers{..} =
-       -- The combo of the gossip set + psq = the whole set of peers
-       availableForGossip
-    <> Set.fromList (PSQ.keys nextGossipTimes)
-    == Map.keysSet allPeers
-
-       -- The gossip set and psq do not overlap
- && Set.null
-      (Set.intersection
-         availableForGossip
-        (Set.fromList (PSQ.keys nextGossipTimes)))
-
```

Comment:
Why this invariant does not hold anymore?

Reply:
This invariant was moved to `EstablishedPeers` module!

Reply:
Right, because we run peer sharing only for established peers.